### DISCUSSION TITLE
How to use alpine functions in fetched modal (Alpine 2.7)?

### DISCUSSION THREAD

**Discussion by QJan84** [alpinejs/alpine#3542] [1. Help] [answered]
How can I use Alpine functions **in the modal** if I get this with `fetch`? 
Every Alpine statement in `/data.php` gives me **8235:3 Uncaught ReferenceError: warengruppen is not defined**.

**Open Modal:** 
`<button @click="$dispatch('modal', {isOpen: true,  url: './data.php?uid=555888'})">Open Modal</button>`

**Build Modal:**
  <div x-data="getModal()">
    <template x-on:modal.window="url = $event.detail.url; isOpen = $event.detail.isOpen; fetchContent();">
    </template>
  
    <template x-if="content">
      <div x-show="isOpen">
  
        <?php /* modal content */ ?>
        <div class="fixed flex justify-center items-center inset-0 w-full h-full z-50">
          <div class="w-full max-w-screen-md relative">
            <div x-html="content">
              
            </div>
          </div>
        </div>
      </div>
    </template>
  </div>
  
**Script:**
```Javascript
  <script>
    // site:: open modal
    function getModal() {
      return {
        isLoading: false,
        isOpen: false,
        url: "",
        content: null,
        size: 'max-w-screen-sm',
        fetchContent() {
          this.isLoading = true;
          fetch(`/dashboard/templates/php/_modals/${this.url}`)
            .then((response) => response.text())
            .then((html) => {
              this.isLoading = false;
              this.content = html;
            })
            .catch((err) => console.log("ERROR", err));
        },
      };
    }
  </script>
```

**Comment by ekwoka** [comment]
What is the content of the modal?

The part you show here all works fine and as expected.

**Comment by QJan84** [comment]
So far the modal works. The PHP for it is loaded via `fetch` from the `data.php` into the modal.

The problem now is that the code that is loaded into the modal from the `data.php` does not recognize any Alpine code/ does not work.

This can be e.g. a list where I can select entries:

Results in an error:
**Uncaught ReferenceError: warengruppen is not defined.**

```PHP
  <div x-data="warengruppen()" x-init="loadOptions()">
    <div>
      <ul>
        <?php // nur freigegebene warengruppen - oben in Liste anzeigen ?>
        <template x-for="(option,index) in options" :key="options[option].value">
          <li>
            <div @click="remove(index, $event)">toggle btn</div>
            <span x-model="option" x-text="option.text"></span>
          </li>
        </template>
        <?php // alle warrengruppen abzüglich der freigegebenen ?>
        <template x-for="(option,index) in options" :key="option">
          <li>
            <div @click="select(index, $event)">toggle btn</div>
            <span x-model="option" x-text="option.text"></span>
          </li>
        </template>
      </ul>
    </div>
  </div>
```

Above code is placed in 
```PHP
  <div x-html="content">
    <!-- fetch content from php file -->
  </div>
```
and does not work.

```Javascript
    function warengruppen() {
      return {
        options: <?php echo json_encode($warengruppen); ?>, // alle warengruppen
        selected: [], // nur freigegebene warengruppen - oben in Liste anzeigen
        filtered: '', // alle warrengruppen abzüglich der freigegebenen
        select(index, event) {
          if (!this.options[index].selected) {
            this.options[index].selected = true;
            this.options[index].element = event.target;
            this.selected.push(index);
          } else {
            this.selected.splice(this.selected.lastIndexOf(index), 1);
            this.options[index].selected = false
          }
        },
        remove(index, option) {
          this.options[option].selected = false;
          this.selected.splice(index, 1);
        },
        loadOptions() {
          const options = document.getElementById('select').options;
          for (let i = 0; i < options.length; i++) {
            this.options.push({
              value: options[i].value,
              text: options[i].innerText,
              selected: options[i].getAttribute('selected') != null ? options[i].getAttribute('selected') : false
            });
          }
        },
        selectedValues(){
          return this.selected.map((option)=>{
            return this.options[option].value;
          })
        }
      };
    }
```

**Reply by ekwoka** [reply]
Is that script being run before you add that html?

Because it needs to.

**Reply by QJan84** [reply]
Nope, but it works in an x-data directive:

            x-data="{
              options: <?php echo $json_string?>,
              selectedOptions: [],
              select(id) {
                var item = this.options.find(item => item.id === id);
                item.active = true;
                this.getSelectedOptions();
              },
              remove(id){
                var item = this.options.find(item => item.id === id);
                item.active = false;
                item.cm = false;
                this.getSelectedOptions();
              },
              toggleCM(id){
                var item = this.options.find(item => item.id === id);
                item.cm = (item.cm) ? false : true;
                this.getSelectedOptions();
              },
              getSelectedOptions(){
                let test = this.options.filter((item) => item.active == true);
                this.selectedOptions = test;
              },
              init(){
                this.getSelectedOptions();
              }
            }"
            x-init="init()"

**Reply by ekwoka** [reply]
tip: `init` is automatically called, you don't need to do `x-init="init"`

But it seems your issue is that you aren't running the code to add to new `data` before trying to init the tree.

So just do that first.

**Comment by QJan84** [comment]
Update:

This works:
```
  <div x-data="{ text: 'Alpine JS'}">
    <span x-text="text"></span>
  </div>
```

This does not work -> **755:3 Uncaught ReferenceError: getText is not defined**:
```
          <div x-data="getText">
            <span x-text="hello"></span>
          </div>
          <script>
            function getText() {
              return {
                hello: 'Hello Alpine JS'
              };
            }
          </script>
```

What can I do to make functions work?

**Reply by SimoTod** [reply, marked as answer]
You can't dynamically inject script tags like that. The browser won't process them for security reasons, the problem is not the alpine code (which runs, in fact it returns an error).

**Reply by ekwoka** [reply]
To add to the above, script tags WILL evaluate most of the time when added as a unique script element.

So the above would work if you did

```js
const script = document.createElement('script')
script.textContent = `function getText() {
              return {
                hello: 'Hello Alpine JS'
              };
            }`
 document.body.append(script)
 
 but anything that adds a block on html, or a whole element tree, will not evaluate, since the browser considered that risky. Adding the element individually does work.
 
 for things like turbo and such, they actually go through and grab all the script tags from the swapped in content and then separately add them to the page to get passed this.
